apps/dotcom/sync-worker/src/UserDataSyncer.ts

```diff
@@ -0,0 +1,303 @@
+	userId: string
+	sequenceId: string
+	bootId: string
+}
+
+export interface ZMutationCommit {
+	type: 'mutation_commit'
+	userId: string
+	mutationNumber: number
+	// if the sequence id changes make sure we still handle the mutation commit
+	sequenceId: string
+}
+export interface ZForceReboot {
+	type: 'force_reboot'
+	sequenceId: string
+}
+
+export type ZReplicationEvent =
+	| ZRowUpdateEvent
+	| ZBootCompleteEvent
+	| ZMutationCommit
+	| ZForceReboot
+
+type BootState =
+	| {
+			type: 'init'
+			promise: PromiseWithResolve
+	  }
+	| {
+			type: 'connecting'
+			bootId: string
+			sequenceId: string
+			promise: PromiseWithResolve
+			bufferedEvents: Array<ZReplicationEvent>
+			didGetBootId: boolean
+			data: null | ZStoreData
+	  }
+	| {
+			type: 'connected'
+			bootId: string
+			sequenceId: string
+	  }
+
+export class UserDataSyncer {
+	state: BootState = {
+		type: 'init',
+		promise: promiseWithResolve(),
+	}
+	replicator: TLPostgresReplicator
+	db: postgres.Sql
+
+	store = new OptimisticAppStore()
+
+	sentry
+	// eslint-disable-next-line local/prefer-class-methods
+	private captureException = (exception: unknown, eventHint?: EventHint) => {
+		// eslint-disable-next-line @typescript-eslint/no-deprecated
+		this.sentry?.captureException(exception, eventHint) as any
+		if (!this.sentry) {
+			console.error(exception)
+		}
+	}
+
+	constructor(
+		private ctx: DurableObjectState,
+		private env: Environment,
+		private userId: string,
+		private broadcast: (message: ZServerSentMessage) => void
+	) {
+		this.sentry = createSentry(ctx, env)
+		this.replicator = env.TL_PG_REPLICATOR.get(
+			env.TL_PG_REPLICATOR.idFromName('0')
+		) as any as TLPostgresReplicator
+		this.db = getPostgres(env)
+		this.reboot(false)
+	}
+
+	private debug(...args: any[]) {
+		// uncomment for dev time debugging
+		// console.log(...args)
+		if (this.sentry) {
+			// eslint-disable-next-line @typescript-eslint/no-deprecated
+			this.sentry.addBreadcrumb({
+				message: args.join(' '),
+			})
+		}
+	}
+
+	private queue = new ExecutionQueue()
+
+	async reboot(delay = true) {
+		// TODO: set up analytics and alerts for this
+		this.debug('rebooting')
+		await this.queue.push(async () => {
+			if (delay) {
+				await sleep(1000)
+			}
+			try {
+				await this.boot()
+			} catch (e) {
+				this.captureException(e)
+				this.reboot()
+			}
+		})
+	}
+
+	private updateStateAfterBootStep() {
+		assert(this.state.type === 'connecting', 'state should be connecting')
+		if (this.state.didGetBootId && this.state.data) {
+			// we got everything, so we can set the state to connected and apply any buffered events
+			assert(this.state.type === 'connecting', 'state should be connecting')
+			const promise = this.state.promise
+			const bufferedEvents = this.state.bufferedEvents
+			const data = this.state.data
+			this.state = {
+				type: 'connected',
+				bootId: this.state.bootId,
+				sequenceId: this.state.sequenceId,
+			}
+			this.store.initialize(data)
```

Comment:
Just noticed that `store.initialize` does not clear optimistic part of the store, but it probably should so that those updates don't stay in the store between reboots?

Reply:
i think what should happen is it should fetch the mutation number during reboots and clear any optimistic updates that happened up to and including that mutation number. I'll add that in this PR.